tests/rc_string_ut/rc_string_ut.c

```diff
@@ -26,40 +42,432 @@ static void on_umock_c_error(UMOCK_C_ERROR_CODE error_code)
+    char* test_string = (char*)test_malloc_func(sizeof(const_test_string));
+    ASSERT_IS_NOT_NULL(test_string);
+
+    (void)memcpy(test_string, const_test_string, sizeof(const_test_string));
+
+    STRICT_EXPECTED_CALL(malloc(IGNORED_ARG));
+
+    umock_c_negative_tests_snapshot();
+
+    for (size_t i = 0; i < umock_c_negative_tests_call_count(); i++)
     {
-        umock_c_negative_tests_deinit();
-        TEST_MUTEX_RELEASE(g_testByTest);
+        if (umock_c_negative_tests_can_call_fail(i))
+        {
+            umock_c_negative_tests_reset();
+            umock_c_negative_tests_fail_call(i);
+
+            // act
+            THANDLE(RC_STRING) rc_string = rc_string_create_with_custom_free(test_string, test_free_func, test_string);
+
+            ///assert
+            ASSERT_IS_NULL(rc_string, "On failed call %zu", i);
+        }
     }
 
+    // cleanup
+    test_free_func(test_string);
+}
+
+/* Tests_SRS_RC_STRING_01_018: [ When the THANDLE(RC_STRING) reference count reaches 0, free_func shall be called with free_func_context to free the memory used by string. ]*/
+TEST_FUNCTION(when_reference_count_reaches_0_the_custom_free_function_is_called)
+{
+    // arrange
+    const char const_test_string[] = "";
+    char* test_string = (char*)test_malloc_func(sizeof(const_test_string));
+    ASSERT_IS_NOT_NULL(test_string);
+
+    (void)memcpy(test_string, const_test_string, sizeof(const_test_string));
+
+    THANDLE(RC_STRING) rc_string = rc_string_create_with_custom_free(test_string, test_free_func, test_string);
+    ASSERT_IS_NOT_NULL(rc_string);
+
+    umock_c_reset_all_calls();
+
+    STRICT_EXPECTED_CALL(test_free_func(test_string));
+    STRICT_EXPECTED_CALL(free(IGNORED_ARG));
+
+    // act
+    THANDLE_DEC_REF(RC_STRING)(rc_string);
+
+    // assert
+    ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+}
+
+/* Tests_SRS_RC_STRING_01_018: [ When the THANDLE(RC_STRING) reference count reaches 0, free_func shall be called with free_func_context to free the memory used by string. ]*/
+TEST_FUNCTION(when_reference_count_reaches_0_the_custom_free_function_is_called_context_NULL)
+{
+    // arrange
+    const char const_test_string[] = "";
+    char* test_string = (char*)test_malloc_func(sizeof(const_test_string));
+    ASSERT_IS_NOT_NULL(test_string);
+
+    (void)memcpy(test_string, const_test_string, sizeof(const_test_string));
+
+    THANDLE(RC_STRING) rc_string = rc_string_create_with_custom_free(test_string, test_free_func_do_nothing, NULL);
+    ASSERT_IS_NOT_NULL(rc_string);
+
+    umock_c_reset_all_calls();
+
+    STRICT_EXPECTED_CALL(test_free_func_do_nothing(NULL));
+    STRICT_EXPECTED_CALL(free(IGNORED_ARG));
+
+    // act
+    THANDLE_DEC_REF(RC_STRING)(rc_string);
+
+    // assert
+    ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+
+    // cleanup
+    test_free_func(test_string);
+}
+
 END_TEST_SUITE(rc_string_unittests)
```

Comment:
>END_TEST_SUITE [](start = 0, length = 14)

I think there are missing tests, where THANDLE_DEC_REF(RC_STRING) is invoked on a string constructed with move semantics.

Those tests should assert that "free" is called and not the test_free_func.

Reply:
You are right me friend. I shall add the test to make sure of what we're calling.

---
In reply to: [460264002](https://github.com/Azure/azure-c-util/pull/52#discussion_r460264002) [](ancestors = 460264002)

Reply:
Added

---
In reply to: [460278823](https://github.com/Azure/azure-c-util/pull/52#discussion_r460278823) [](ancestors = 460278823,460264002)